ambry-api/src/main/java/com.github.ambry/messageformat/BlobProperties.java

```diff
@@ -32,6 +32,7 @@
   private final boolean isEncrypted;
   private long blobSize;
   private long timeToLiveInSeconds;
+  private short recordVersion;
```

Comment:
Since both Put record and Update record need `version`, can we add `version` in a common place? `MessageHeader_Format` looks a good place for me.

Can we name it to a `special` name to avoid confusion with record version or format version? `updateVersion` or `evolutionVersion`?

Reply:
I agree that `recordVersion` may not be the best name.  `updateVersion` I think may be clearer so I'm open to changing to that.

I understand your motivation for wanting to put it in the header due to its being shared by both PUT blobs and UPDATE blobs, but I don't think it's the right place as the purpose of the header is simply to tell the receiving parser the payload size and field demarcations of the payload.  Plugging in `updateVersion` into the message header feels like subverting the purpose of the message header, and adding identification logic in a place that it doesn't belong.  Also TTLUpdate doesn't use `updateVersion`

Reply:
BlobProperties is for user, we shouldn't include `updateVersion` to confuse them.

The message header is a metadata container. `updateVersion` is also a kind of metadata to all messages. Plus, It's shared by all message.

Adding version identification logic based on message header is what we want. For example, in `BlobStoreRecover.recover()`, we need `updateVersion` to recover `Index` which should also include a `updateVersion`. With version in message header, all message can share same code and recover with easy logic. Particularly, you don't need to parse BlobProperties for PUT. 
 
I understand ttlUpdate doesn't need a version, but we should add version to it. Reasons are:
1. logic easy, just look one step back and increase version number, for all kind of message.
2. open the door of multiple ttlUpdate.

Reply:
Does that mean you're ok with the current code or do you still think having it in the message header is preferable?  Did you happen to look at how indexing, compaction, and/or replication are done at a deeper level and see if one way would make implementing those other needed functions easier?

Reply:
Prefer message header. 

Looking at index, compaction, replication. Will send out a doc by EOW.

Reply:
Have we reached an agreement on 1. where the version should be placed. 2. explicit rules of updating the versions ? According to original Priyesh's design, it seems we use convention something like P0, D0, U0, D1, U1 ...